### ISSUE TITLE
goog.provide somehow prevents the compiler from checking goog.require statements

### ISSUE THREAD

**Issue by aproxs** [bazelbuild/rules_closure#38]
I'm using bazel 0.2.2, installed following the instructions in bazel.io

Repro:
- Create a new folder
- sync the sources for bazelbuild/rules_closure
- cd rules_closure/closure/compiler/test

edit hello.js and add the following line to the top of the file:

```
goog.require('foo');
```

Now verify that the build behaves as expected (it should fail)

```
bazel build :hello_bin
```

Now add the following line to the top of hello.js:

```
goog.provide('bar');
```

Build again:

```
bazel build :hello_bin
```

The build should have failed, however the library just builds fine.

Just to rule out a bug in the compiler I tried to repro using the closure compiler service here: http://closure-compiler.appspot.com (see https://developers.google.com/closure/compiler/docs/gettingstarted_ui for deatils about it) and it works as expected.

This behavior seems related to the "--depencency_mode=LOOSE" flag int he compiler.

I might be missing something important here but: why does the closure_js_binary rule specify that flag by default if I don't specify a 'main' value in the rule configuration?

**Comment by jart** [comment]
> The build should have failed, however the library just builds fine.

As of #49, library rules will now also fail to build if dependencies are missing in this way.

> [Why is --dependency_mode=LOOSE specified by default.]

Note: You may have noticed that the `depmode` functionality mentioned in the README isn't implemented yet. I'll fix that soon. That behavior is going to redesigned somewhat. So it's currently not possible to set --dependency_mode=NONE.

LOOSE is the default because it allows the Closure Compiler to strip out JS files that aren't being used. For example, when you add //closure/library to your deps list, 800+ JS files get schlepped in. You want the compiler to get rid of the ones whose namespaces aren't being required. NONE is only really useful for JS code with side-effects that doesn't use any sort of namespace management.

Tangential: Speaking of which, I honestly can't even find the Closure Compiler code that implements the --dependency_mode flag. How on earth.

So what exactly is the problem you're experiencing?

**Comment by aproxs** [comment]
Awesome, so if #49 fixes the issue that should be it.

What I was experiencing: I was configuring my build to use closure rules and the closure library, and I knew that I had a significant amount of compilation errors in the code.
In one of the first configurations that I tried, I 'bazel build' my project and there were 0 issues reported, so that's why I was suspicious and found this.

The way I worked around it was by specifying a "main" symbol, which triggers the use of a --dependency_mode that surfaces errors properly.

Thanks for taking a look at it!
